esp_idf_panic_decoder/pc_address_decoder.py

```diff
@@ -23,50 +34,185 @@ def __init__(
-            if b'?? ??:0' not in translation:
-                decoded = translation.decode()
-                return decoded if not is_rom else decoded.replace('at ??:?', 'in ROM')
+            batch_output = subprocess.check_output(cmd, cwd='.')
         except OSError as err:
             red_print(f'{" ".join(cmd)}: {err}')
+            return {}
         except subprocess.CalledProcessError as err:
             red_print(f'{" ".join(cmd)}: {err}')
             red_print('ELF file is missing or has changed, the build folder was probably modified.')
-        return None
+            return {}
+
+        decoded_output = batch_output.decode(errors='ignore')
+
+        return PcAddressDecoder.parse_addr2line_output(decoded_output, is_rom=is_rom)
+
+    @staticmethod
+    def parse_addr2line_output(
+        output: str,
+        is_rom: bool = False,
+    ) -> Dict[str, List[PcAddressLocation]]:
+        """
+        Parse the output of addr2line.
+        :param output: The output of addr2line as a string.
+        :param is_rom: If True, replace '??' paths with 'ROM' as paths are not available from ROM ELF files.
+        :return: Map from each address to a list of its source locations (with multiple indicating an inlined function).
+        """
+
+        # == addr2line output example ==
+        # 0xabcd1234  # Aad # First input address
+        # foo()       # A0f # Function
+        # foo.c:123   # A0p # Source location
+        # 0x1234abcd  # Bad # Second input address
+        # inlined()   # B0f # Inlined function
+        # bar.c:456   # B0p # Source location
+        # bar()       # B1f # Function which inlined inlined()
+        # bar.c:789   # B1p # Source location
+        # ...         # ... # ... more entries
+
+        # Step 1: Split into sections representing each address and its trace (A**, B**)
+        sections = re.split(ADDR2LINE_ADDRESS_LOOKAHEAD_RE, output)
+
+        result: Dict[str, List[PcAddressLocation]] = {}
+        for section in sections:
+            section = section.strip() # Remove trailing newline
+            if not section:
+                continue
+
+            # Step 2: Split the section by newlines (Aad, A0f, A0p)
+            lines = section.split('\n')
+
+            # Step 3: First line is the address (Aad)
+            address = lines[0].strip()
+
+            # Step 4: Build trace by consuming lines in pairs (A0f + A0p)
+            #         Multiple entries indicate inlined functions (B0f + B0p, B1f + B1p, etc.)
+            trace: List[PcAddressLocation] = []
+            for func, path_line in zip(map(str.strip, lines[1::2]), map(str.strip, lines[2::2])):
+                path_match = ADDR2LINE_FILE_LINE_RE.match(path_line)
+                path = path_match.group('file') if path_match else path_line
+                line = path_match.group('line') if path_match else ''
+                if path == '??' and is_rom:
+                    path = 'ROM'
```

Comment:
```suggestion
                if func == '??' and path == '??':
                    continue
                if path == '??' and is_rom:
                    path = 'ROM'
```

@peterdragun How about this? No point returning it from here either if there is no useful information.

Reply:
That could work, but it would require some additional handling, as this results in the following output, which is not expected:
![image](https://github.com/user-attachments/assets/3bb9cbef-57e2-4b38-9c18-c40df081a7bb)

Reply:
Added `valid` variable to this loop - if any entry for a given address contains a function or path we keep the entire trace, otherwise if none of the entries do then we discard the address.

Probably overkill and not sure if this is even possible _but_ this will ensure that if there was ever a situation like this we would keep the full trace, instead of suggesting that `0x40078000` == func at file.c:

```
0x40078000: ?? at ??:0
(inlined by) func at file.c:1
(inlined by) ?? at ??:0